subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java

```diff
@@ -466,7 +467,14 @@ public void ensureLegacySnapshottingInputsClosed() {
         @Override
         public void validate(WorkValidationContext validationContext) {
             FileOperations fileOperations = ((ProjectInternal) task.getProject()).getFileOperations();
-            context.getTaskProperties().validate(new DefaultTaskValidationContext(fileOperations, reservedFileSystemLocationRegistry, validationContext));
+            Class<?> taskType = GeneratedSubclasses.unpackType(task);
+            // TODO This should probably use the task class info store
+            boolean cacheable = taskType.isAnnotationPresent(CacheableTask.class);
```

Comment:
Should this check if validation was activated via the runtime API as well? We already add a `cacheIf` spec if caching has been enabled via the annotation, right? Can we re-use `shouldDisableCaching`?

Reply:
I've been thinking about this, but it's complicated. We currently do validation _before_ we determine the cacheability of the work item is determined. `shouldDisableCaching()` is called from there.

What we can do is split validation of cache-related rules to be executed later in the execution pipeline. I think we should do that eventually, but not now.

We should probably skip cachable-only checks if cacheability was _disabled_ via the runtime API, too. And we could run a few checks on properties registered via the runtime API. There's a bunch of things we can add here.

I think what we have now is a good first step, though.

Reply:
We said that we wanted to disable caching if we find validation problems related to cacheable tasks. So maybe the order to first validate and then check for cacheabiiity could work out well.

I am fine with leaving as is for now.